include/seqan3/io/sequence/sequence_file_format_fastq.hpp

```diff
@@ -0,0 +1,326 @@
+            }
+        }
+
+        /* 2nd ID line */
+        if (*stream_it != '+') // [[unlikely]]
+        {
+            throw parse_error{std::string{"Expected '+' on beginning of 2nd ID line, got: "} +
+                              detail::make_printable(*stream_it)};
+        }
+        detail::consume(stream_view | view::take_line_or_throw);
+
+        /* Qualities */
+        auto qview = stream_view | ranges::view::remove_if(is_space)                  // this consumes trailing newline
+                                 | ranges::view::take_exactly(sequence_size_after - sequence_size_before);//TODO orthrow
+        if constexpr (seq_qual)
+        {
+            // seq_qual field implies that they are the same variable
+            assert(std::addressof(sequence) == std::addressof(qualities));
+            ranges::copy(qview | view::char_to<typename value_type_t<qual_type>::quality_alphabet_type>,
+                         ranges::begin(qualities) + sequence_size_before);
+        }
+        else if constexpr (!detail::decays_to_ignore_v<qual_type>)
+        {
+            ranges::copy(qview | view::char_to<value_type_t<qual_type>>,
+                         ranges::back_inserter(qualities));
+        }
+        else
+        {
+            detail::consume(qview);
+        }
+
+        // make sure "buffer at end" implies "stream at end"
+        if ((std::istreambuf_iterator<char>{stream} == std::istreambuf_iterator<char>{}) &&
+            (!stream.eof()))
+        {
+            stream.get(); // triggers error in stream and sets eof
+        }
+    }
+
+    //!\copydoc sequence_file_out_format_concept::write
+    template <typename stream_type,     // constraints checked by file
+              typename seq_type,        // other constraints checked inside function
+              typename id_type,
+              typename qual_type>
+    void write(stream_type                     & stream,
+               sequence_file_out_options const & options,
+               seq_type                       && sequence,
+               id_type                        && id,
+               qual_type                      && qualities)
+    {
+        if constexpr (sized_range_concept<seq_type> && sized_range_concept<qual_type>)
+        {
+            assert(ranges::size(sequence) == ranges::size(qualities));
+        }
+
+        ranges::ostreambuf_iterator stream_it{stream};
+
+        // ID
+        if constexpr (detail::decays_to_ignore_v<id_type>)
+        {
+            throw std::logic_error{"The ID field may not be set to ignore when writing FASTQ files."};
+        }
+        else
+        {
+            if (ranges::empty(id)) //[[unlikely]]
+                throw std::runtime_error{"The ID field may not be empty when writing FASTQ files."};
+
+            stream_it = '@';
+            ranges::copy(id, stream_it);
+
+            if (options.add_carriage_return)
+                stream_it = '\r';
+
+            stream_it = '\n';
```

Comment:
Shouldn't we define a shortcut, e.g. `add_eol` or `endl`, as we need this multiple times in every file format?